src/rust/cryptography-key-parsing/src/spki.rs

```diff
@@ -0,0 +1,137 @@
+// This file is dual licensed under the terms of the Apache License, Version
+// 2.0, and the BSD License. See the LICENSE file in the root of this repository
+// for complete details.
+
+use cryptography_x509::common::{AlgorithmParameters, EcParameters, SubjectPublicKeyInfo};
+
+use crate::{KeyParsingError, KeyParsingResult};
+
+pub fn parse_public_key(
+    data: &[u8],
+) -> KeyParsingResult<openssl::pkey::PKey<openssl::pkey::Public>> {
+    let k = asn1::parse_single::<SubjectPublicKeyInfo>(data)?;
+
+    match k.algorithm.params {
+        AlgorithmParameters::Ec(ec_params) => match ec_params {
+            EcParameters::NamedCurve(curve_oid) => {
+                let curve_nid = match curve_oid {
+                    cryptography_x509::oid::EC_SECP192R1 => openssl::nid::Nid::X9_62_PRIME192V1,
+                    cryptography_x509::oid::EC_SECP224R1 => openssl::nid::Nid::SECP224R1,
+                    cryptography_x509::oid::EC_SECP256R1 => openssl::nid::Nid::X9_62_PRIME256V1,
+                    cryptography_x509::oid::EC_SECP384R1 => openssl::nid::Nid::SECP384R1,
+                    cryptography_x509::oid::EC_SECP521R1 => openssl::nid::Nid::SECP521R1,
+
+                    cryptography_x509::oid::EC_SECP256K1 => openssl::nid::Nid::SECP256K1,
+
+                    cryptography_x509::oid::EC_SECT233R1 => openssl::nid::Nid::SECT233R1,
+                    cryptography_x509::oid::EC_SECT283R1 => openssl::nid::Nid::SECT283R1,
+                    cryptography_x509::oid::EC_SECT409R1 => openssl::nid::Nid::SECT409R1,
+                    cryptography_x509::oid::EC_SECT571R1 => openssl::nid::Nid::SECT571R1,
+
+                    cryptography_x509::oid::EC_SECT163R2 => openssl::nid::Nid::SECT163R2,
+
+                    cryptography_x509::oid::EC_SECT163K1 => openssl::nid::Nid::SECT163K1,
+                    cryptography_x509::oid::EC_SECT233K1 => openssl::nid::Nid::SECT233K1,
+                    cryptography_x509::oid::EC_SECT283K1 => openssl::nid::Nid::SECT283K1,
+                    cryptography_x509::oid::EC_SECT409K1 => openssl::nid::Nid::SECT409K1,
+                    cryptography_x509::oid::EC_SECT571K1 => openssl::nid::Nid::SECT571K1,
+
+                    #[cfg(not(any(CRYPTOGRAPHY_IS_LIBRESSL, CRYPTOGRAPHY_IS_BORINGSSL)))]
```

Comment:
Does libressl really not support brainpool?

Reply:
Hmm, now I'm not sure. The same branch exists for them in ec.rs. Let me do a separate PR removing that.

Reply:
Ugh, it's a rust-openssl issue. Let me fix.

Reply:
https://github.com/sfackler/rust-openssl/pull/2150